[NFS Undercover] Add dynamic shadow resolution control and cascade scale control

xan1242 opened the pull request:
It's not much but it allows for nicer shadows by bruteforce.

- Added injection points for controlling the resolution of the `CSM::pAtlasTexture`
- Added a cave for the X resolution multiplier (for shadow cascades)
- Added a cave for scaling the `CSM::fScaleLerp` value - increasing this increases draw distance, but hurts the quality

xan1242 commented:
There is also an array `CSM::fCSMScaleLevel[]` at `0x00D53F38` which has 3 values for controlling the scale of each cascade scale individually. If you scale all of the cascades at the same time too far, they can overdraw back into the nearest cascade which causes bugs.

I will try to put controls for those as well.

xan1242 commented:
I've added the control values for `CSM::fCSMScaleLevel[]` to allow for more granular control and removed the `CSM::fScaleLerp` scaling.

This MIGHT break cutscenes as they have their own values, but from quick testing I see nothing out of the ordinary and should be OK.

HarGabt commented:
NFSUC shadow upscale? Unbelievable!
Looking forward to this!

Looking through the code related to shadow resolution got me wondering: if `5460` is the maximum value because of the game using 3 shadow cascades, what if it's possible to force the game use only one of the cascades and then upscale to its maximum `16384`?

xan1242 commented:
Technically, you can do it already by setting the ShadowLevel setting to 1, but that clearly removes a lot from the render. (Also I will have to add this into consideration. If you use the low setting, you should be able to create a 16k texture)

The shader itself is reconfigurable as well, with the same settings (off, low, medium, high) as noted inside its source code.

In order to find this, I need to know all of the places in the code where it asks for the shadow detail setting (via DALManager, setting 0xE). I found one place but that messed everything up when I forced it to "high" mode (but also restored missing stuff funnily).

In any case, I'm researching both variants (shader and single shadow).

ThirteenAG commented:
Can you make default ini with "nice" settings, and also write default values in the comment? I think it's better that way, since there's quite a few options.

xan1242 commented:
Sure!

I just need a good default resolution since 5460 (16k) causes lag on some weaker hardware.

It's very blocky even at that res so I'm not sure what to do yet. I'll test some settings out and find nicer defaults over stock that aren't too demanding.

HarGabt commented:
>Technically, you can do it already by setting the ShadowLevel setting to 1

As of now, I can't do that - the script still narrows `16384` down to `5460` even with `ShadowLevel` 1

And what I meant by forcing one of the cascades is that some cascades don't cast shadows on cars and some parts of the world on PC version. 
So, IMO it would be better either to: 
- give an option to force the game to use the closest cascade only (that casts shadows on everything) and increase its scale
- or to dig deeper and find the piece of code that has to do with casting shadows on objects and force cascades to cast shadows on _all of the objects_.

HarGabt commented:
Follow-up
<details> <summary>Pictures behind the spoiler</summary>
<h3> ShadowLevel 1</h3> - Only mid cascade is rendered. 

![Безымянный](https://user-images.githubusercontent.com/44652400/188416407-ced32d74-785b-4a39-8fd4-d85c789694f1.png)

<h3> ShadowLevel 2</h3> - Mid and far cascade rendered. Even though it may seem that shadows are cast on car, there aren't any.

![Безымянный2](https://user-images.githubusercontent.com/44652400/188416418-fda0e944-7d93-4228-8fb5-6e0c0f53c2e5.png)

<h3> ShadowLevel 3</h3> All three cascades rendered. 

![Безымянный3](https://user-images.githubusercontent.com/44652400/188416425-b18247cf-1d7f-4878-b6b9-dbde4fcb4152.png)
</details>

xan1242 commented:
> > Technically, you can do it already by setting the ShadowLevel setting to 1
> 
> As of now, I can't do that - the script still narrows `16384` down to `5460` even with `ShadowLevel` 1
> 
> And what I meant by forcing one of the cascades is that some cascades don't cast shadows on cars and some parts of the world on PC version. So, IMO it would be better either to:
> 
> * give an option to force the game to use the closest cascade only (that casts shadows on everything) and increase its scale
> * or to dig deeper and find the piece of code that has to do with casting shadows on objects and force cascades to cast shadows on _all of the objects_.

Both can be done. As a matter of fact, I've already done this.

The issue with 1st point is that there is no point in doing this because it scales the stencil out regardless, losing the quality.

2nd point I did do. Something then scales it down 3x on the width and goes all out of place. Not sure what to do here exactly. It seems to be handled by the shader (as it adds a "_Low" string at the end of technique name) so I can't exactly figure this one out as easily.

Shaders being in BBFX format doesn't help, so that's going to need some time to figure out.

xan1242 commented:
I've implemented dynamic ShadowsRes scaling according to the number of cascades you set in settings.

Lower you go in Shadow quality, it lowers the number of cascades, but increases the highest possible res (yes, I know it's counter-intuitive, but blame Black Box for that).

Now you can go with up to 1x16k, 2x8192 or 3x5461.

If the setting in the ini exceeds any of these, it'll automatically divide it to the maximum possible value.

@ThirteenAG I will polish the ini with nice values soon, I have to play with this a bit more before I can decide what qualifies as nice...

xan1242 commented:
So, I will need your input on this (@AeroWidescreen you can join in too):

Stock
![image](https://user-images.githubusercontent.com/8014093/189037182-f910b3fe-6813-4f52-8802-ccbc4012a5ef.png)


ShadowsRes = 5461, CSMScaleNear = 100.0, CSMScaleMid = 175.0, CSMScaleFar = 300.0
![image](https://user-images.githubusercontent.com/8014093/189036250-6ee35fcc-7889-4991-97b3-ef70fe665018.png)

As you can see, the quality drop is pretty significant, but inverse to the draw distance.

This place might be a bit of a bad example, so let's head on to a tunnel.

Stock
![image](https://user-images.githubusercontent.com/8014093/189037791-11788129-aa5f-4aff-986f-89b93cf560a9.png)

ShadowsRes = 5461, CSMScaleNear = 100.0, CSMScaleMid = 175.0, CSMScaleFar = 300.0
![image](https://user-images.githubusercontent.com/8014093/189037811-f6533f9a-2436-4341-a366-2c8a603842e0.png)


Currently it has to be a compromise between quality and draw distance due to scaling. Anything lower than 100.0 on CSMScaleNear is really noticeable in player view's draw distance. (This value affects the car's shadow most notably)

xan1242 commented:
> ![Безымянный3](https://user-images.githubusercontent.com/44652400/188416425-b18247cf-1d7f-4878-b6b9-dbde4fcb4152.png)

BTW how did you get the view of the texture/stencil at the bottom? I could use this...

HarGabt commented:
@xan1242 
I used your [reshade fork](https://github.com/xan1242/reshade) and [MeshEdges.fx](https://github.com/Daodan317081/reshade-shaders/blob/master/Shaders/MeshEdges.fx) applied to shadow depth buffer

xan1242 commented:
@HarGabt How exactly? What did you edit in MeshEdges.fx to get it working? Can you elaborate on everything please?
 Right now it doesn't want to even read the depth whatsoever for me...

ThirteenAG reviewed (changes requested):
```
[SHADOWS]
ShadowsRes = 1024						 // Controls the resolution of dynamic shadows. Shadow map contains up to 3 cascades on the X axis. Resulting resolution will be ShadowsRes / g_ShadowEnable. (1024 = Default | 16384 = Max)
CSMScale = 1.0							 // Overall scale of the cascade shadow map. Larger the number, greater the draw distance, but the quality is lower. The same applies to the values below. (1.0 = Default)
CSMScaleNear = 5.0						 // Nearest cascade (5.0 = Default)
CSMScaleMid = 30.0						 // Mid cascade (30.0 = Default)
CSMScaleFar = 170.0						 // Far cascade (170.0 = Default)
```

xan1242 commented:
If there is a way to scale/transform this, it'd be awesome.

Basically, this is a single cascade:
![image](https://user-images.githubusercontent.com/8014093/189069715-4c5c0ed2-4192-4f6d-b29e-c588d4c2603f.png)

This is a 3x cascade:
![image](https://user-images.githubusercontent.com/8014093/189070377-bea2e081-1ccb-47ed-a305-03a465654205.png)

The display scale of this texture is dictated in the function at 0x00780070. Nopping 0x007800AF causes the display scaling to be disabled, making it expect what is in the 2nd image. So when it draws something it's all messed up.

![image](https://user-images.githubusercontent.com/8014093/189071425-b7558c3e-c9a5-42e7-8129-d71933ef9702.png)

So if we scale it the same as the 2nd image, we should see a single cascaded render with all details (hopefully).

(Also I just realized I should treat Y res separately from X... whoops! Will try that next! Should result in better looking shadows potentially...)

xan1242 commented:
Yet another thing I found during researching for scaling... Adding it in ASAP.

Basically, there seem to be some exclusion flags in `RenderShadowView()` at 0x0079E31D. Setting them to -1 causes more things to be rendered to the shadow stencil. I see no drawbacks to this, so I've added it in.

Before
![image](https://user-images.githubusercontent.com/8014093/189088290-e3b0b590-c6de-4a65-84e8-51ec303bc14b.png)

After
![image](https://user-images.githubusercontent.com/8014093/189088378-91d43563-db11-42d8-b05b-4a93cde07382.png)

xan1242 commented:
Ahhhh... FE is broken sadly. Fixing it now.

![image](https://user-images.githubusercontent.com/8014093/189090800-e1383222-c045-4456-8083-1fac883fe5cf.png)

xan1242 commented:
Undoing it for now... It's too broken to fix at the moment. It's good to know for the future if we decide to hack this part for shadows.

Unless there's anything else you guys have to add or suggest, I'd call this PR complete.

If & when we figure out the single cascaded rendering, that should be a separate thing IMO.

xan1242 commented:
The thing I was talking about previously. 

It looks less blocky now but slightly dithered. Should Y resolution be separately configurable?

16384x5461
![image](https://user-images.githubusercontent.com/8014093/189036250-6ee35fcc-7889-4991-97b3-ef70fe665018.png)

16384x16384 3x Cascades
![image](https://user-images.githubusercontent.com/8014093/189128537-dba2a24d-c833-4a00-8d6f-b5f53b28bf3d.png)

HarGabt commented:
In what ways or terms (apart from FrontEnd) does shadow LOD setting break graphics?

If we need to workaround the FE issue, we might want to use a check if the player is in frontend [as seen in Aero_'s NFSPS HD Reflections](https://github.com/AeroWidescreen/NFSHDReflections/blob/6c00fa3ad8c110fa41f0921384012d6ecd7667e7/SOURCE/NFSPS/assembly.h#L48)

Or not?

xan1242 commented:
There's more to that. It doesn't just break FE (I could just check the game state to fix it), but it also causes certain shadows to actually disappear, so it isn't ideal without digging deeper into it.

I'd leave that feature for a separate PR just to keep things clean. We can figure that one out later.

xan1242 commented:
Updated the resolution calculations to use floating point numbers, as 3x cascades aren't nicely divisible.

Also updated the ini defaults to use 8192, so the result will be 16384x8192 as default on 3x cascades. If there are any user complaints to excessive lagging or whatever similar, then we'll need to address it by reducing the res and CSM scale back...

ca5dx commented:
For some reason, using the nfs steam .exe that is supposedly the most mod friendly (yet breaks semi trucks spawning with trailers, which I mentioned on another ticket) I can't go above 4096 shadow resolution in the generic fix .ini. The default number of 8192 just makes the game crash on boot.
